src/utils/ThreadPool.h

```diff
@@ -0,0 +1,384 @@
+                for (T i = chunk_first ; i < chunk_last ; ++i)
+                {
+                    shared_state.loop_body(i);
+                }
+                th_lock.lock();
+                if (--shared_state.chunks_remaining == 0)
+                {
+                    shared_state.work_done.notify_one();
+                }
+            });
+    }
+
+    // Do work while parallel_for's tasks are running
+    thread_pool->work_while(lock, [&]{ return shared_state.chunks_remaining > 0; });
+    while(shared_state.chunks_remaining > 0) // Wait until all the task are completed
+    {
+        shared_state.work_done.wait(lock);
+    }
+}
+
+/*!
+ *  \brief An implementation of parallel for.
+ *  Overload for iterating over containers with random access iterators.
+ */
+template<typename Container, typename F>
+auto parallel_for(Container& container, F&& loop_body, size_t chunk_size_factor=1, size_t chunks_per_worker=8)
+  -> std::void_t<decltype(container.end() - container.begin())>
+{
+    parallel_for(container.begin(), container.end(), std::forward<F>(loop_body), chunk_size_factor, chunks_per_worker);
+}
+
+
+//! \private Internal state for run_multiple_producers_ordered_consumer()
+template<typename Producer, typename Consumer> class MultipleProducersOrderedConsumer;
+
+/*!
+ * \brief Runs parallel producers and buffers the results to be consumed serially in indices order.
+ *
+ * Producers run while there is available space in the buffer.
+ * Only 0 or 1 consumer runs at any time.
+ *
+ * When a thread produces the item waited for consumption, it turns itself into a consumer.
+ * When the consumer thread completes an item, it notifies an eventual producer waiting on a full buffer.
+ * When the consumer thread encounter an item not yet produced, it becomes a producer again.
+ *
+ * Produced items are stored into a shared ring buffer.
+ * The item type must be nullable in order to differentiate free (not yet produced) slots.
+ *
+ * \param fist,last The numerical range of elements to produce.
+ * \param producer Given an index/iterator, produces a non-null value of a nullable type (eg pointer, optional, etc...).
+ * \param consumer Consumes an item produced by `producer`.
+ * \param max_pending_per_worker Number of allocated slots per worker for items waiting to be consumed.
+ */
+template<typename P, typename C>
+void run_multiple_producers_ordered_consumer(ssize_t first, ssize_t last, P&& producer, C&& consumer, size_t max_pending_per_worker=8)
+{
+    ThreadPool* thread_pool = Application::getInstance().thread_pool;
+    assert(thread_pool);
+    assert(max_pending_per_worker > 0);
+    const size_t max_pending = max_pending_per_worker * (thread_pool->thread_count() + 1);
+    MultipleProducersOrderedConsumer(first, last, std::forward<P>(producer), std::forward<C>(consumer), max_pending).run(*thread_pool);
+}
+
+template<typename Producer, typename Consumer>
+class MultipleProducersOrderedConsumer
```

Comment:
This class stores user closures and state variables together such that they can referenced by spawned tasks through a single reference. `parallel_for` does the same but with a scoped struct having no methods.

State for `run_multiple_producers_ordered_consumer` is a bit more involved and I found that it's easier to use a full fledged class. On the other hand, this class should not be used directly as I don't want to pretend that there is other lifecycles than constructor->`run()`->destructor.